pkg/tkctl/cmd/dump/dump.go

```diff
@@ -0,0 +1,429 @@
+		return err
+	}
+
+	podList, err := o.kubeCli.CoreV1().Pods(o.namespace).List(metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s,%s in (%s)", label.InstanceLabelKey, tc.Name, label.ComponentLabelKey,
+			strings.Join([]string{label.TiDBLabelVal, label.TiKVLabelVal, label.PDLabelVal}, ",")),
+	})
+	if err != nil {
+		return err
+	}
+
+	if _, err := rWriter.Write([]byte("----------------pods---------------\n")); err != nil {
+		return err
+	}
+
+	// dump detail information and logs of pods.
+	for _, pod := range podList.Items {
+		if err := NewPodDumper(o.kubeCli, pod, int64(o.since.Seconds()), o.byteReadLimit).Dump(o.logPath, rWriter); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// tidbClusterDumper generates information about a tidbclusters object.
+type tidbClusterDumper struct {
+	tc *v1alpha1.TidbCluster
+}
+
+// NewTiDBClusterDummper returns a tidbClusterDumper.
+func NewTiDBClusterDumper(tc *v1alpha1.TidbCluster) *tidbClusterDumper {
+	return &tidbClusterDumper{
+		tc: tc,
+	}
+}
+
+// Dump dump the details of tidbclusters from a particular namespace.
+func (d *tidbClusterDumper) Dump(logPath string, resourceWriter io.Writer) error {
+	logFile, err := os.Create(filepath.Join(logPath, fmt.Sprintf("%s-%s-tidbcluster-info.yaml", d.tc.Name, d.tc.Namespace)))
+	if err != nil {
+		return err
+	}
+	defer logFile.Close()
+
+	if _, err := resourceWriter.Write([]byte("----------------tidbclusters---------------\n")); err != nil {
+		return err
+	}
+
+	if err = DumpObj(d.tc, resourceWriter); err != nil {
+		return err
+	}
+
+	data, err := yaml.Marshal(d.tc)
+	if err != nil {
+		return err
+	}
+
+	return writeString(logFile, string(data))
+}
+
+// tidbClusterStatefulDumper generates information about a statefulset by a specify tidbcluster.
+type tidbClusterStatefulDumper struct {
+	kubeCli *kubernetes.Clientset
+	tc      *v1alpha1.TidbCluster
+}
+
+// NewTiDBClusterStatefulDumper returns a tidbClusterStatefulDumper
+func NewTiDBClusterStatefulDumper(tc *v1alpha1.TidbCluster, kubeCli *kubernetes.Clientset) *tidbClusterStatefulDumper {
+	return &tidbClusterStatefulDumper{
+		tc:      tc,
+		kubeCli: kubeCli,
+	}
+}
+
+// Dump dump the details of statefulsets from a particular tidb cluster.
+func (d *tidbClusterStatefulDumper) Dump(logPath string, resourceWriter io.Writer) error {
+	logFile, err := os.Create(filepath.Join(logPath, fmt.Sprintf("%s-%s-statefulsets-info.yaml", d.tc.Name, d.tc.Namespace)))
+	if err != nil {
+		return err
+	}
+	defer logFile.Close()
+
+	if _, err := resourceWriter.Write([]byte("----------------statefulset---------------\n")); err != nil {
+		return err
+	}
+
+	for _, sn := range []string{
+		controller.TiDBMemberName(d.tc.Name),
+		controller.TiKVMemberName(d.tc.Name),
+		controller.PDMemberName(d.tc.Name),
+	} {
+		ps, err := d.kubeCli.AppsV1().StatefulSets(d.tc.Namespace).Get(sn, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		if err = writeString(logFile, "#"+sn+"\n"); err != nil {
```

Comment:
the exported YAML file doesn't have `apiVersion` and `kind` attribute.

Reply:
Fixed.